src/Microsoft.TemplateEngine.Cli/CommandParsing/NewCommandInputCli2.cs

```diff
@@ -0,0 +1,305 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.DotNet.Cli.CommandLine;
+using Microsoft.TemplateEngine.Abstractions;
+using Microsoft.TemplateEngine.Edge.Template;
+
+namespace Microsoft.TemplateEngine.Cli.CommandParsing
+{
+    public class NewCommandInputCli2 : INewCommandInput
+    {
+        private ParseResult _parseResult;
+        private IReadOnlyList<string> _args;
+        private string _templateNameArg;
+
+        private Command _currentCommand;
+        private Func<Task<int>> _invoke;
+
+        private IReadOnlyDictionary<string, string> _templateParamValues;
+        private IReadOnlyDictionary<string, IList<string>> _templateParamCanonicalToVariantMap;
+
+        // used for parsing args outside the context of a specific template.
+        private readonly Command _noTemplateCommand;
+        private readonly string _commandName;
+
+        public NewCommandInputCli2(string commandName)
+        {
+            _commandName = commandName;
+            _noTemplateCommand = CommandParserSupport.CreateNewCommandWithoutTemplateInfo(_commandName);
+            _currentCommand = _noTemplateCommand;
+        }
+
+        public bool HasParseError
+        {
+            get
+            {
+                return _parseResult.Errors.Count() > 0;
```

Comment:
Can't seem to find the definition of `ParseResult`, can we do this as `.Count` or `.Any()` if it's an IEnumerable?